log.go

```diff
@@ -193,6 +225,38 @@ func (this *SrvLog) SetLogsEnabled(name string, val bool) {
 	logs.enabled = val
 }
 
+// WithField stores a key, value pair in the logger that will be included in all subsquent logs.
+func (this *SrvLog) WithField(key string, value interface{}) *SrvLog {
+	return this.WithFields(map[string]interface{}{
+		key: value,
+	})
+}
+
+// WithFields stores all the given key, value pairs in the logger that will be includeded in all subsequent logs.
+func (this *SrvLog) WithFields(fields map[string]interface{}) *SrvLog {
+	this.lock.RLock()
+	defer this.lock.RUnlock()
+
+	newSubs := make(map[string]*LogService)
+	for key, val := range this.subs {
+		newSubs[key] = val
+	}
+
+	newFields := make(map[string]interface{})
+	for key, val := range this.fields {
+		newFields[key] = val
+	}
+	for key, val := range fields {
+		newFields[key] = val
+	}
+
+	derived := &SrvLog{
+		subs:   newSubs,
+		fields: newFields,
+	}
+	return derived
+}
+
```

Comment:
I wasn't sure how this was supposed to work without an example, but I think I've got it now. I believe you are intending the returned and modified copy of the SrvLog object to be used as a temporary and then let it get cleaned up as it goes out of scope? Did you choose to do it this way because you foresee overwriting the same key as you go deeper into the stack and you want to make sure to support reverting back to previous values as you unwind? If that wasn't a goal, why not a register / defer unregister pattern on the primary SrvLog object?

Reply:
Ah yeah. The intent is that a copy of the SrvLog is created that is a superset of the original (i.e. with different fields) and that the original logger is unmodified. I did it that way for a couple of reasons:

1. I wanted the caller to be able to not have to worry about forgetting to revert the changes. Using object lifetime scope seemed a natural choice.
2. I was mirroring the API style seen in logrus, and to a lesser degree zap and zerolog. Those libraries are much more sophisticated in their implementations though - in particular zerolog which pulls off the scoping without any allocations and superior performance numbers.